test/blackbox-tests/test-cases/dune-project-meta/github11205.t

```diff
@@ -0,0 +1,66 @@
+Support for negation in package dependencies:
+
+  $ cat > dune-project <<EOF
+  > (lang dune 3.18)
+  > (generate_opam_files)
+  > (package
+  >   (name foo)
+  >   (allow_empty)
+  >   (depends
+  >    (ocp-indent
+  >     (not
+  >       (or
+  >        (and
+  >         (not :with-test)
+  >         (>= 1.0))
+  >        (not
+  >         (and
+  >          (not (= :os win32))
+  >          (not (>= 1.5)))))))
+  >    (not (>= 2.0))))
+  > EOF
+  $ dune build
+  $ cat foo.opam
+  # This file is generated by dune, edit dune-project instead
+  opam-version: "2.0"
+  depends: [
+    "dune" {>= "3.18"}
+    "ocp-indent" {!(!with-test & >= "1.0" | !(!os = "win32" & !>= "1.5"))}
+    "not" {>= "2.0"}
+    "odoc" {with-doc}
+  ]
+  build: [
+    ["dune" "subst"] {dev}
+    [
+      "dune"
+      "build"
+      "-p"
+      name
+      "-j"
+      jobs
+      "@install"
+      "@runtest" {with-test}
+      "@doc" {with-doc}
+    ]
+  ]
+  x-maintenance-intent: ["(latest)"]
+
+Not supported before 3.18:
+
+  $ cat > dune-project <<EOF
+  > (lang dune 3.17)
+  > (generate_opam_files)
+  > (package
+  >   (name foo)
+  >   (allow_empty)
+  >   (depends
+  >    (ocp-indent
+  >     (not :with-test))))
+  > EOF
+  $ dune build
+  File "dune-project", line 8, characters 4-20:
+  8 |     (not :with-test))))
+          ^^^^^^^^^^^^^^^^
+  Error: Not operator is only available since version 3.18 of the dune
+  language. Please update your dune-project file to have (lang dune 3.18).
+  [1]
```

Comment:
I think it would be also good to show that `dune pkg lock` correctly handles `not` and will avoid adding a package that has a `not` constraint (and error out if there is no solution).

Probably not in this cram test but in the dune pkg cram tests.

Reply:
Good point, I added some negation tests for `dune pkg lock` :)